Added support for external libraries

vinay-lanka opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The PR follows [our contributing guidelines](https://arduino.github.io/arduino-cli/CONTRIBUTING/#pull-requests)
- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls) before creating one)
- [x] Tests for the changes have been added (for bug fixes / features)


* **What kind of change does this PR introduce?**
   #### Feature - 
   Added support for external libraries -
   - added `--git-url` flag to add libraries via git url
     Usage : `arduino-cli lib install <Library Name> --git-url <url>`
   - added `--zip-path` flag to add libraries by pointing path to zip file
     Usage : `arduino-cli lib install <Library Name> --zip-path <absolute path>`


* **What is the new behavior?**
One can add external libraries and existing libraries right from the git repositories.
Can also include libraries from other vendors distributed as a zip file.


* **Other information**:
Have not added tests for zip-path flag. (Suggestions on how to download and locate a zip library with pytest)

---
See [how to contribute](https://arduino.github.io/arduino-cli/CONTRIBUTING/)

vinay-lanka commented:
Is there anything I could do to improve this change? 
I feel this could be a slight improvement to the whole user experience as discussed in #128

altjz commented:
Any progress on this feature?

silvanocerza commented:
Hey @vinay-lankam, sorry for the late response on this PR! 
It would be great if you have the time to rebase it on top of `master` to resolve the current conflicts.
Don't worry if you don't have the time I'll find another way.

vinay-lanka commented:
Hey, 
Sure, could I have a day or two to rebase it to the `master`? 
I'll try to resolve all the current conflicts ASAP.

silvanocerza commented:
That's perfect, thanks!

vinay-lanka commented:
Hey, @silvanocerza, I'm done with resolving all conflicts.

silvanocerza commented:
@vinay-lanka thanks! I'll do a review as soon as I can, probably tomorrow.

silvanocerza commented on arduino/libraries/librariesmanager/install.go:
```diff
@@ -85,3 +91,99 @@ func (lm *LibrariesManager) Uninstall(lib *libraries.Library) error {
+}
+
+//Unzip takes the ZipLibPath and Extracts it to LibsDir
+func Unzip(src string, dest string) error {
```
There's no need for this function to be public, it's used only inside `InstallZipLib`.
Also you can avoid repeating the type of the arguments, `src, dest string` is the preferred way.

vinay-lanka replied:
Yeah sure, I'll do the same with Fetch()?
It also is called only once in InstallGitLib.

silvanocerza commented on arduino/libraries/librariesmanager/install.go:
```diff
@@ -85,3 +91,99 @@ func (lm *LibrariesManager) Uninstall(lib *libraries.Library) error {
+		// Store filename/path for returning and using later on
+		fpath := filepath.Join(dest, f.Name)
+
+		// Check for ZipSlip. More Info: http://bit.ly/2MsjAWE
```
Please don't abbreviate URLs, use the original one.

silvanocerza commented on arduino/libraries/librariesmanager/install.go:
```diff
@@ -85,3 +91,99 @@ func (lm *LibrariesManager) Uninstall(lib *libraries.Library) error {
+}
+
+//InstallGitLib  installs a library hosted on a git repository on the specified path.
+func (lm *LibrariesManager) InstallGitLib(Name string, gitURL string) error {
```
Same as above `name, gitURL string`, notice also `name` is not capitalized.

silvanocerza commented on arduino/libraries/librariesmanager/install.go:
```diff
@@ -85,3 +91,99 @@ func (lm *LibrariesManager) Uninstall(lib *libraries.Library) error {
+}
+
+//Fetch Clones the repository to LibsDir
+func Fetch(name string, url string, dest string) error {
```
Much like `Unzip` this function can also be private, also the arguments type thing.

silvanocerza commented on arduino/libraries/librariesmanager/install.go:
```diff
@@ -85,3 +91,99 @@ func (lm *LibrariesManager) Uninstall(lib *libraries.Library) error {
+
+//Fetch Clones the repository to LibsDir
+func Fetch(name string, url string, dest string) error {
+	directory := dest + "/" + name
```
Doesn't this break on Windows?

silvanocerza commented on cli/lib/install.go:
```diff
@@ -40,65 +40,92 @@ func initInstallCommand() *cobra.Command {
+			os.Exit(errorcodes.ErrGeneric)
+		}
+	} else if installFlags.gitURL != "" {
+		GitlibraryInstallReq := &rpc.GitLibraryInstallReq{
```
Don't capitalize variable names, same above for `ZiplibraryInstallReq`.

silvanocerza commented on cli/lib/install.go:
```diff
@@ -40,65 +40,92 @@ func initInstallCommand() *cobra.Command {
+	} else if installFlags.gitURL != "" {
+		GitlibraryInstallReq := &rpc.GitLibraryInstallReq{
+			Instance: instance,
+			Name:     args[0],
```
Why `args[0]` as `Name`? Can extract it directly from the `gitURL` inside the `GitLibraryInstall` function?

vinay-lanka replied:
So I thought it'd be better if we followed the same syntax of `arduino-cli lib install LIBRARY[@VERSION_NUMBER](S) [flags]`. Even though it seems redundant, I thought it would minimize confusion as the same command would have different arguments for different flags.

silvanocerza replied:
Ah! I see why you did it but I think it would be better to avoid the redundancy, also if we go that way we'd have to do it for the installation suing the zip file too.

silvanocerza commented on commands/lib/install.go:
```diff
@@ -75,3 +75,44 @@ func installLibrary(lm *librariesmanager.LibrariesManager, libRelease *libraries
+func ZipLibraryInstall(ctx context.Context, req *rpc.ZipLibraryInstallReq) (*rpc.ZipLibraryInstallResp, error) {
+	res := &rpc.ZipLibraryInstallResp{}
+	lm := commands.GetLibraryManager(req.GetInstance().GetId())
+	Path := req.GetPath()
```
Don't capitalize variable names.

silvanocerza commented on commands/lib/install.go:
```diff
@@ -75,3 +75,44 @@ func installLibrary(lm *librariesmanager.LibrariesManager, libRelease *libraries
+	res := &rpc.ZipLibraryInstallResp{}
+	lm := commands.GetLibraryManager(req.GetInstance().GetId())
+	Path := req.GetPath()
+	if err := installZipLibrary(lm, Path); err != nil {
```
We don't need the `installZipLibrary`, call directly `LibrariesManager.InstallZipLib`.

vinay-lanka replied:
The normal Library install procedure is divided into two functions `LibraryInstall()` and `installLibrary()` so I thought I'd follow the same style. As it does not actually have a unique function, I'll remove it asap.

silvanocerza replied:
Interesting, the normal procedure will probably need to be changed to if it's like that, not in this PR though.

silvanocerza commented on commands/lib/install.go:
```diff
@@ -75,3 +75,44 @@ func installLibrary(lm *librariesmanager.LibrariesManager, libRelease *libraries
+	res := &rpc.GitLibraryInstallResp{}
+	lm := commands.GetLibraryManager(req.GetInstance().GetId())
+	URL := req.GetUrl()
+	Name := req.GetName()
```
Capitalized variable name.

silvanocerza commented on commands/lib/install.go:
```diff
@@ -75,3 +75,44 @@ func installLibrary(lm *librariesmanager.LibrariesManager, libRelease *libraries
+	lm := commands.GetLibraryManager(req.GetInstance().GetId())
+	URL := req.GetUrl()
+	Name := req.GetName()
+	if err := installGitLibrary(lm, Name, URL); err != nil {
```
We don't need `installGitLibrary` either, call directly the `LibrariesManager` function.

silvanocerza commented on rpc/commands/lib.proto:
```diff
@@ -294,3 +294,22 @@ enum LibraryLocation {
+
+message GitLibraryInstallReq {
+    Instance instance = 1;
+    string name = 2;
```
As I said in a comment above we probably won't need the `name` since it can be extracted from the URL.

silvanocerza commented on test/test_lib.py:
```diff
@@ -152,6 +152,18 @@ def test_install(run_command):
     assert "Error resolving dependencies for MD_Parola@3.2.0: dependency 'MD_MAX72xx' is not available" in result.stderr
 
 
+def test_install_flags(run_command):
```
Could you write a test for the zip installation too?

vinay-lanka replied:
Yeah sure, I just figured I can `wget` a zipped library and test the install.

silvanocerza replied:
Best to use the `lib download` command and then install that downloaded one, also I don't think `wget` is available on Windows.

To find the downloaded library you'll probably need the `downloads_dir` fixture for that test.

silvanocerza commented:
@vinay-lanka please let me know when the PR is ready for review again, or if there are other issues so we can see what to do.

vinay-lanka commented:
Hey, I was waiting for #1049 to get merged because I had a problem with the integration tests. I rebased after that but I'm still getting an assertion error at  `test/test_core.py::test_core_search - assert 2 == 0` . I'm trying to figure it out but the rest is ready for a review.

silvanocerza commented:
Don't worry about that test failing for now, I'm aware of it and trying to understand what's going on exactly.

silvanocerza commented:
Those tests have been fixed, should I review it now?

vinay-lanka commented:
I've cleaned up the code (Sorry for the numerous errors!)
I've added tests for the zip path based install and added a test for `lib download` as well

silvanocerza commented on arduino/libraries/librariesmanager/install.go:
```diff
@@ -85,3 +91,84 @@ func (lm *LibrariesManager) Uninstall(lib *libraries.Library) error {
+		// Store filename/path for returning and using later on
+		fpath := filepath.Join(libsDir.String(), f.Name)
+
+		// Check for ZipSlip.
+		if !strings.HasPrefix(fpath, filepath.Clean(libsDir.String())+string(os.PathSeparator)) {
+			return fmt.Errorf("%s: illegal file path", fpath)
+		}
+
+		filenames = append(filenames, fpath)
+
+		if f.FileInfo().IsDir() {
+			// Make Folder
+			os.MkdirAll(fpath, os.ModePerm)
+			continue
+		}
+
+		// Make File
+		if err = os.MkdirAll(filepath.Dir(fpath), os.ModePerm); err != nil {
+			return err
+		}
+
+		outFile, err := os.OpenFile(fpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
+		if err != nil {
+			return err
+		}
+
+		rc, err := f.Open()
+		if err != nil {
+			return err
+		}
+
+		_, err = io.Copy(outFile, rc)
+
+		// Close the file without defer to close before next iteration of loop
+		outFile.Close()
+		rc.Close()
+
+		if err != nil {
+			return err
+		}
```
Would be great if you could use the [codeclysm/extract](github.com/codeclysm/extract) library, it's already around the CLI, see `arduino/resources/install.go` for example.

silvanocerza commented on arduino/libraries/librariesmanager/install.go:
```diff
@@ -85,3 +91,84 @@ func (lm *LibrariesManager) Uninstall(lib *libraries.Library) error {
+	}
+	i := strings.LastIndex(url, "/")
+	folder := strings.TrimRight(url[i+1:], ".git")
+	path := libsDir.String() + "/" + folder
```
Concatenate using `Path.Join()` like so `libsDir.Join(folder)`, using the path separator explicitly like so might break on some OS.

silvanocerza commented on cli/lib/install.go:
```diff
@@ -36,69 +36,99 @@ func initInstallCommand() *cobra.Command {
+					os.Exit(errorcodes.ErrGeneric)
+				}
+				for _, dep := range depsResp.GetDependencies() {
+					feedback.Printf("%s depends on %s@%s", libRef, dep.GetName(), dep.GetVersionRequired())
+					if existingDep, has := toInstall[dep.GetName()]; has {
+						if existingDep.GetVersionRequired() != dep.GetVersionRequired() {
+							// TODO: make a better error
+							feedback.Errorf("The library %s is required in two different versions: %s and %s",
+								dep.GetName(), dep.GetVersionRequired(), existingDep.GetVersionRequired())
+							os.Exit(errorcodes.ErrGeneric)
+						}
 					}
+					toInstall[dep.GetName()] = dep
 				}
-				toInstall[dep.GetName()] = dep
 			}
 		}
-	}
 
-	for _, library := range toInstall {
-		libraryInstallReq := &rpc.LibraryInstallReq{
-			Instance: instance,
-			Name:     library.Name,
-			Version:  library.VersionRequired,
-		}
-		err := lib.LibraryInstall(context.Background(), libraryInstallReq, output.ProgressBar(), output.TaskProgress())
-		if err != nil {
-			feedback.Errorf("Error installing %s: %v", library, err)
-			os.Exit(errorcodes.ErrGeneric)
+		for _, library := range toInstall {
+			libraryInstallReq := &rpc.LibraryInstallReq{
+				Instance: instance,
+				Name:     library.Name,
+				Version:  library.VersionRequired,
+			}
+			err := lib.LibraryInstall(context.Background(), libraryInstallReq, output.ProgressBar(), output.TaskProgress())
+			if err != nil {
+				feedback.Errorf("Error installing %s: %v", library, err)
+				os.Exit(errorcodes.ErrGeneric)
+			}
```
I'm not sure setting the minimum number of args to `0` is the best approach here.
Let's try this thing, change the minimum number of args back to `1`, the flags `git-url` and `zip-path` to boolean flags. Doing so we can call the CLI like so: 
```
arduino-cli lib install <my_lib_name>
arduino-cli lib install --git-url <my_git_url>
arduino-cli lib install --zip-path <my_zip_file>
```
You'll have to change a bit the logic in this file but I think it's a good improvement and will make the command clearer to use.

silvanocerza commented on commands/lib/install.go:
```diff
@@ -75,3 +75,29 @@ func installLibrary(lm *librariesmanager.LibrariesManager, libRelease *libraries
+	lm := commands.GetLibraryManager(req.GetInstance().GetId())
+	url := req.GetUrl()
+	if err := lm.InstallGitLib(url); err != nil {
+		res.Status = "Error installing Git Library"
+		return res, err
```
If there's an error don't return res, just the error. `return nil, err` is fine, same for `ZipLibraryInstall`.

silvanocerza commented:
Approved, looks great now! :partying_face: 
Don't worry about the `verify-links` failing, we're aware of it and is not an issue.
Thank you very much for the contribution, we greatly appreciate the help of the community and look forward to future PRs. :)
I'll merge this as soon all checks finish.

silvanocerza commented:
Merged! Thanks again @vinay-lanka.

vinay-lanka commented:
Thank you so much for the opportunity! 
I learnt a lot while contributing so would love to open PRs in the future! :)

ubidefeo commented:
@vinay-lanka PRs like these are always welcome 👍
